### PR TITLE
[None][feat] Add startup self-benchmarking with batched prefill

### PR DESCRIPTION
**Pull request by sachalmalick**
## Description

This carries for the startup self-benchmarking implementation and adds a prefill batch-size dimension, keeping TensorRT-LLM aligned with the related Dynamo/vLLM work in https://github.com/ai-dynamo/dynamo/pull/7779.

The startup benchmark covers prefill input length, prefill batch size, cached-KV reads, decode context length, and decode batch size. Prefill points are capped by both the scheduler sequence limit and token budget. Every batch member receives deterministic per-rank request metadata and a distinct cache salt, and a point is recorded only after telemetry observes the full batch. The change preserves TP-rank lockstep, KV-read seed/measurement pairing, output-schema behavior, and disaggregated context/generation startup.

The PR also adds the nested `SelfBenchmarkConfig` surface, serving documentation, API-stability updates, CI enrollment, and focused unit coverage for the complete startup self-benchmarking feature.

This supersedes #16191, which had gone stale against `main`. The work is unchanged in substance — it has been squashed into a single commit and rebased onto current `main`. See **Rebase notes** below.

## Rebase notes

Rebased from the old merge base onto `main` at `ddd62066bf` (1304 commits of drift). Twenty conflict hunks across nine files were resolved; all were adjacent-addition conflicts rather than semantic collisions. Two required an ordering decision:

- **`py_executor.py` / `_prepare_and_schedule_batch`** — the benchmark's `observe_scheduled_batch` + drain block now runs *before* main's new `_pad_empty_attention_dp_batch`, so the benchmark observes the capacity scheduler's real verdict rather than a padded dummy batch.
- **`py_executor.py` / `_schedule`** — main's new `prepare_expect_snapshot_points` runs against the full `self.active_requests`; the benchmark's `should_hold_from_scheduler` filter then builds the local `active_requests` list handed to the scheduler.

`test_py_executor_creator_mla_cache_reuse_sync.py` also needed a manual merge: the new self-benchmark test had been auto-aligned against four tests main added in the same region. All of main's tests are preserved, and `_run_create_py_executor` carries the union of both kwarg sets.

The resulting net diff is identical in size to #16191 (5431 insertions, 35 deletions across 17 files), and every line added by the original PR is present apart from two import-rewrap artifacts — `TrtLlmArgs` and `BuildConfig`, which `main` has since removed from `tensorrt_llm/llmapi/__init__.py` outright.

## Validation

The results below were produced on the pre-rebase tree (#16191). **They have not been re-run since the rebase** — please treat CI on this PR as the authoritative signal for the rebased code.

| Scope | Result |
|---|---|
| Repository checks | The complete PR-diff pre-commit suite passed. The focused self-benchmark, serve-config parsing, and LLM API-stability suite passed **64 tests in 8.42 seconds**. |
| Single-GPU runtime | Qwen3-0.6B on **1× L40S** completed prefill points `(ISL, batch) = (1,1), (1,4), (64,1)`. The corresponding `(numContextRequests, numCtxTokens)` telemetry was exactly `(1,1), (4,4), (1,64)`. |
| Tensor parallel | **TP=2 on 2× L40S** completed successfully on both ranks. Both reported `status=complete`, `valid=true`, `world_size=2`, and `tp_size=2`, with identical prefill points including batch 4. A normal **2-token generation** also completed after the startup sweep. |
| Disaggregated serving | **TP=2 context + TP=2 generation on 4× L40S** passed end-to-end. Context, generation, and proxy health checks returned **HTTP 200**. Context ranks recorded prefill batch 4 as 4 requests/4 tokens at ISL 1; generation ranks recorded batch 4 with `numGenRequests=4` at context lengths 1 and 63. NIXL over UCX initialized on every worker rank, and the proxy returned an **8-token OpenAI completion**. |

On the tested node, NCCL's P2P/CUMEM path also hung in a minimal one-float control all-reduce, independent of TensorRT-LLM and this PR. Setting `NCCL_P2P_DISABLE=1` selected the shared-memory transport; the raw two-rank control then returned the expected reduced value **3.0**, and both TP=2 and disaggregated validations passed.

## PR Checklist

- [x] Description explains the motivation and implementation.
- [x] Tests cover the new code paths.
- [x] Additive API change is labeled `api-compatible`.
- [x] Documentation and API-stability references are updated.
- [x] No new dependencies or CODEOWNERS changes.

## Dev Engineer Review

- Added startup self-benchmarking for the PyTorch executor.
- Added `SelfBenchmarkConfig` with benchmark modes, granularity, warmup, and output-path settings.
- Added validation for supported decoder-generation and pure tensor-parallel configurations.
- Added batched prefill, cache reuse, decode measurement, and disaggregated context/generation support.
- Added deterministic request metadata, cache salts, full-batch telemetry validation, TP-rank consensus, and atomic schema-v2 JSON artifacts.
- Added CUDA-graph-aware batch sampling with eager-tail handling and TP-rank capture-set intersection.
- Disabled self-benchmarking during temporary KV-cache estimation.
- Updated serving documentation, API exports, API-stability references, and usage manifests.
- Configuration and API updates are consistent across `llm_args`, executor creation, serving configuration, and API references.
- The implementation affects scheduling, shutdown, draining, timeout handling, and resource cleanup. These paths require focused review.
- The A10 test-list entries use the existing YAML format and add five distinct tests. CBTS coverage data was not provided. CI enrollment requires follow-up if coverage validation is required.

## QA Engineer Review

- Added `tests/unittest/_torch/executor/test_self_benchmark.py` coverage for request construction, case generation, scheduler planning, TP-rank consensus, lifecycle transitions, draining, interruption, error handling, KV-cache reuse, response processing, artifact validation, and CUDA-graph-aware sampling.
- Added `test_kv_estimation_keeps_args_stable_and_gates_self_benchmark`.
- Added request utility coverage for disabled broadcasting, benchmark-marker filtering, and merged requests.
- Updated executor creator, disaggregated benchmark, and dual-pool KV-cache test fixtures.
- Added `SelfBenchmarkConfig` coverage for default output paths, YAML merging, defaults, and iteration statistics.
- Added `unittest/_torch/executor/test_self_benchmark.py` to `tests/integration/test_lists/test-db/l0_a10.yml`.
- `unittest/llmapi/test_llm_args.py` is covered by both `l0_a10.yml` and `l0_cpu.yml`.
- The modified request utility, executor creator, disaggregated benchmark, and dual-pool KV-cache tests are not explicitly listed in the checked CI test lists.
- Verdict: needs follow-up.

### CODE REVIEW

**Review by nvpohanh — COMMENTED** [review]
approved for kv-cache manager

**Review by nvpohanh — APPROVED** [review]
approved for kv-cache manager

File: tensorrt_llm/llmapi/llm_args.py [hunk 1]
```diff
@@ -3089,6 +3092,61 @@ def validate_histogram_buckets(cls, v: Optional[List[float]],
...
+        default=5,
+        description=
+        "Number of synthetic warmup iterations to run before collecting "
+        "self-benchmark results.",
+        status="prototype")
+
+    output_path: str = Field(
+        default=_DEFAULT_SELF_BENCHMARK_OUTPUT_PATH,
+        description="Path to write self-benchmark results JSON.",
+        status="prototype")
```
[the remark this answers is not in the document]
Comment (BowenFu) [reply tensorrt_llm/llmapi/llm_args.py:3151]: +1

**Review by tburt-nv — APPROVED** [review]

**Review by mzweilz — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/self_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,2437 @@
...
+        return request_id in trial.completed_request_ids
+
+    def request_interrupt(self) -> None:
+        if not self.active:
+            return
+        if not self._interrupt_requested:
+            logger.info("Self-benchmark stopping early because shutdown was requested.")
+        self._interrupt_requested = True
+
+    def local_outcome(self) -> BenchmarkOutcome:
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/self_benchmark.py:650]: `local_outcome` can return CONTINUE indefinitely. If a trial's requests are never admitted or never released (e.g. a resource leak in a new path this sweep exercises at startup), the benchmark stays `active` forever and, because `_fetch_and_enqueue_requests` parks every normal request while active, the server accepts traffic and serves none of it with no error. Add a per-trial iteration cap (or wall-clock budget) that transitions to SKIP with a `trial_timeout` reason, and an overall sweep budget that transitions to ABORT.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 3]
```diff
@@ -3797,6 +3837,16 @@ def _prepare_and_schedule_batch(self):
         scheduled_batch, scheduler_fitting_disagg_gen_init_requests, _ = self._schedule(
         )
 
+        if benchmark is not None and benchmark.active:
+            benchmark.observe_scheduled_batch(scheduled_batch)
+            self._sync_self_benchmark_outcome()
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:3842]: This is the second blocking `tp_allreduce` in the same iteration (the first is before `_schedule`). Under TP both collectives land inside the window whose `hostStepTimeMS` this benchmark records, so small-batch decode points measure MPI latency the production loop never pays. Consider a single exchange per iteration: the pre-schedule call only needs to propagate an interrupt, which could be folded into this one.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 4]
```diff
@@ -2416,6 +2428,22 @@ def _process_iter_stats(
                                         batch_state.scheduled_requests,
                                         micro_batch_id,
                                         batch_state.scheduled_batch_stats)
+        benchmark = self.self_benchmark
+        if benchmark is not None and benchmark.active:
+            stats_dict = json.loads(stats.to_json_str())
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:2433]: `stats.to_json_str()` followed by `json.loads` runs on the hot path of every benchmark iteration. For overlap-scheduler decode points where host time is the critical path this adds measurable per-step overhead to the very numbers being recorded. Reading the fields you need directly from the `IterationStats` object (or deferring serialization until the trial finalizes) would keep the bookkeeping out of the measured window.

File: tensorrt_llm/llmapi/llm_args.py [hunk 5]
```diff
@@ -6114,6 +6185,58 @@ def validate_early_first_token_response(self) -> 'TorchLlmArgs':
...
+                "parallelism yet (moe_ep_size/moe_cluster_size > 1). Only pure "
+                "tensor parallelism is supported.")
+        # Any remaining world_size > 1 must be pure TP (tp_size == world_size).
+        if parallel.world_size != parallel.tp_size:
+            raise ValueError(
+                "self_benchmark_config only supports pure tensor parallelism; "
+                f"got world_size={parallel.world_size} with "
+                f"tp_size={parallel.tp_size}. Disable other forms of "
+                "parallelism to use startup self-benchmarking.")
+        self.enable_iter_perf_stats = True
```
Comment (cascade812) [inline tensorrt_llm/llmapi/llm_args.py:6237]: Forcing `enable_iter_perf_stats = True` here is a permanent side effect: after the sweep finishes the executor keeps computing and queueing iteration stats every step for the life of the process, which the user did not ask for. Either capture the user's value and restore it when the benchmark reaches a terminal state, or document in the field description that enabling self-benchmarking also enables iteration perf stats.

File: tensorrt_llm/llmapi/llm_args.py [hunk 6]
```diff
@@ -3089,6 +3092,65 @@ def validate_histogram_buckets(cls, v: Optional[List[float]],
...
+        status="prototype")
+
+    warmup_iterations: NonNegativeInt = Field(
+        default=5,
+        description=
+        "Number of synthetic warmup iterations to run before collecting "
+        "self-benchmark results.",
+        status="prototype")
+
+    output_path: str = Field(
```
Comment (cascade812) [inline tensorrt_llm/llmapi/llm_args.py:3148]: The default output path is a fixed file under the system temp dir, and `_invalidate_output` unlinks whatever is there at init. Two servers on the same host with default settings (a common test-bench setup, and the disagg ctx/gen pair when co-located) will delete and overwrite each other's results. Suggest including `run_id` or pid in the default filename, or making `output_path` required when the config is present.

File: tensorrt_llm/llmapi/llm_args.py [hunk 7]
```diff
@@ -3089,6 +3092,65 @@ def validate_histogram_buckets(cls, v: Optional[List[float]],
...
+
+    decode_batch_granularity: PositiveInt = Field(
+        default=6,
+        description=
+        "Maximum number of batch size sample points for each decode context length. "
+        "Sample points are snapped onto CUDA graph capture sizes when CUDA graphs are "
+        "enabled, so fewer points may be emitted.",
+        status="prototype")
+
+    warmup_iterations: NonNegativeInt = Field(
```
Comment (cascade812) [inline tensorrt_llm/llmapi/llm_args.py:3141]: Every measured point is a single iteration: a trial's requests finish after one step and `iteration_stats` holds one entry. Only warmup is repeatable. Given the downstream planner fits a model to these numbers, an `iterations_per_case` (or `repeats`) knob would let users trade startup time for variance; at minimum the docs should say each point is a single sample.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 8]
```diff
@@ -5586,18 +5745,48 @@ def _fetch_and_enqueue_requests(self, waiting_queue: WaitingQueue,
...
+                # (accumulate them) and forward only control/special items.
+                # Synthetic benchmark PREFILL requests are NOT injected here:
+                # they are built per-rank in _fetch_and_activate_new_requests so
+                # every TP rank deterministically produces the same forward
+                # batch. Injecting them into rank 0's broadcast queue would
+                # double-inject on non-rank-0 ranks (once via broadcast, once
+                # locally) and desync the per-rank state machine.
+                for req_item in queued_requests:
+                    if req_item.is_normal_request:
+                        self.request_accumulated.append(req_item)
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:5785]: User requests deferred here are invisible: no log, no count in the artifact, and their clients simply wait. Log once when the first request is deferred (with the number of remaining cases) and record the deferred count in the terminal artifact so a slow-start report can be diagnosed.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 9]
```diff
@@ -4354,6 +4497,15 @@ def _executor_loop(self):
                     self._send_kv_async(scheduled_batch.all_requests())
 
                     self._handle_canceled_requests()
+                    benchmark_active = benchmark is not None and benchmark.active
+                    attn_metadata = getattr(self.model_engine, 'attn_metadata',
+                                            None)
+                    kv_cache_dtype_byte_size = getattr(
+                        self.model_engine, 'kv_cache_dtype_byte_size', None)
+                    if benchmark_active:
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:4505]: The `update_resources` / `_handle_responses` order flips when the benchmark is active, in both this loop and `_process_previous_batch`, but neither site says why. If the reason is that `_handle_responses` terminates benchmark requests before their KV bookkeeping runs, please state that in a comment; the next person touching either loop will otherwise see two paths that differ for no visible reason.

File: tensorrt_llm/_torch/pyexecutor/self_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,2437 @@
...
+        )
+        return request
+
+    def _exclude_last_generation_logits(self) -> bool:
+        # Engine-wide flag derived from disable_overlap_scheduler and pp_size;
+        # identical across TP ranks, so it does not perturb per-rank lockstep.
+        return bool(getattr(self._executor, "should_exclude_last_generation_logits", False))
+
+    def _mark_benchmark_request(self, request: "LlmRequest", trial_id: int) -> None:
+        request.is_self_benchmark_request = True
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/self_benchmark.py:1804]: `is_self_benchmark_request` is set dynamically and read via `getattr(..., False)` at eight sites in `py_executor.py`, while the sibling attribute uses the `py_` prefix. Declare `py_is_self_benchmark_request = False` (and `py_self_benchmark_trial_id = None`) in `LlmRequest.__init__` so the attribute is discoverable, consistently named, and the defensive `getattr` calls can become plain attribute reads.

**Review by Mgluhovskoi — APPROVED** [review]

File: docs/source/commands/trtllm-serve/trtllm-serve.rst
Comment (BowenFu) [inline docs/source/commands/trtllm-serve/trtllm-serve.rst]: Thanks for putting this together! The 1.3 release branch is about to be cut from main (planned for 9/29), and until then we're asking larger changes and refactors like this one to wait so main stays stable for the branch-out. If this really needs to make 1.3 or a nightly/weekly release, could you explain why in the PR description and in your review request? I'll resolve this thread once that's there, or after branch-out either way. This is only a timing hold, not a concern with the change itself.
